server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -85,20 +86,13 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             .expandToAllNodes()) {
             return YES_AUTO_EXPAND_ALL;
         }
-        if (node.node() != null) {
-            assert Strings.hasLength(node.node().getHostAddress()) : node;
-            for (RoutingNode checkNode : allocation.routingNodes()) {
-                if (checkNode.node() == null) {
-                    continue;
-                }
-                // check if its on the same host as the one we want to allocate to
-                assert Strings.hasLength(checkNode.node().getHostAddress()) : checkNode;
-                if (checkNode.node().getHostAddress().equals(node.node().getHostAddress())) {
-                    for (ShardRouting assignedShard : assignedShards) {
-                        if (checkNode.nodeId().equals(assignedShard.currentNodeId())) {
-                            return allocation.debugDecision() ? debugNoAlreadyAllocatedToHost(node, allocation) : Decision.NO;
-                        }
-                    }
+        Hosts hosts = allocation.getHosts();
+        Host host = hosts.getHost(node.nodeId());
+        // return yes if can not find the host info of candidate node(no host name & address)
+        if (node.node() != null && host != null) {
+            for (ShardRouting assignedShard : assignedShards) {
+                if (hosts.getHost(assignedShard.currentNodeId()) == host) {
```

Comment:
I wonder if we even need the hosts structure now that we always have a host address and only check for that? Can we do something like:

```
allocation.nodes().getDataNodes().get(assignedShard.currentNodeId()).getHostAddress().equals(node.node().getHostAddress())
```

With your follow-up PR for host name, I would expect it to become so that we only check either the address or the name, never both, which should make the same possible for the host name case?

Reply:
You are right. In this PR, we only need to confirm whether the host of assignedShard is consistent with the host of the current node. The hosts data structure was originally designed by me to implement a total_shards_per_host decider. It's not necessary here.